### ISSUE TITLE
Query with `index IN (subquery)` in subqueries is very slow.

### ISSUE THREAD

**Issue by StarWix** [ClickHouse/ClickHouse#2345]
Example:

```sql
CREATE TABLE test
ENGINE = MergeTree(Partition, x, 8192) AS
SELECT 
    toDate(number % 32768) AS Partition, 
    number AS x
FROM system.numbers 
LIMIT 10000010;

// 0 rows in set. Elapsed: 8.218 sec. Processed 10.03 million rows, 80.22 MB (1.22 million rows/s., 9.76 MB/s.)

// FIRST
SELECT 
    1
FROM test 
WHERE x IN 
(
    SELECT 
        x
    FROM test 
    ORDER BY x ASC
    LIMIT 10000000, 1
);

// 1 rows in set. Elapsed: 3.045 sec. Processed 5.94 million rows, 47.56 MB (1.95 million rows/s., 15.62 MB/s.) 

//SECOND
SELECT 1 FROM (
    SELECT 
        1
    FROM test 
    WHERE x IN 
    (
        SELECT 
            x
        FROM test 
        ORDER BY x ASC
        LIMIT 10000000, 1
    )
);

// 1 rows in set. Elapsed: 10.283 sec. Processed 5.94 million rows, 47.56 MB (578.14 thousand rows/s., 4.63 MB/s.)

//THIRD
SELECT 1 FROM (
    SELECT 1 FROM (
        SELECT 
            1
        FROM test 
        WHERE x IN 
        (
            SELECT 
                x
            FROM test 
            ORDER BY x ASC
            LIMIT 10000000, 1
        )
    )
);

// 1 rows in set. Elapsed: 23.681 sec. Processed 5.94 million rows, 47.56 MB (251.05 thousand rows/s., 2.01 MB/s.)
```

I expect that the last two queries should executed so fast as the first query.

Maybe related to #2301

**Comment by agaurav** [comment]
@alexey-milovidov it has been labelled `st-accepted`. 
has it been already taken ?

**Comment by alexey-milovidov** [comment]
This issue is in queue. The reason is slow index analysis.

**Comment by CurtizJ** [comment]
Analysis of index is not significant in these queries.
This case was optimized in several steps. Firstly there was implemented optimization of sorting with using order key, but difference between queries still existed. Finally performance of nested subqueries was fixed in #9235. This PR eliminates unnecessery reading of table for several times by passing proper `QueryOptions` to subqueries.

Now we have on master:
```
SELECT version()

┌─version()───┐
│ 20.3.1.2532 │
└─────────────┘

1 rows in set. Elapsed: 0.004 sec. 

SET max_partitions_per_insert_block = 50000

Ok.

0 rows in set. Elapsed: 0.002 sec. 

CREATE TABLE test
ENGINE = MergeTree(Partition, x, 8192) AS
SELECT 
    toDate(number % 32768) AS Partition, 
    number AS x
FROM system.numbers
LIMIT 10000010

Ok.

0 rows in set. Elapsed: 4.091 sec. Processed 10.03 million rows, 80.22 MB (2.45 million rows/s., 19.61 MB/s.) 

SELECT 1
FROM test
WHERE x IN 
(
    SELECT x
    FROM test
    ORDER BY x ASC
    LIMIT 10000000, 1
)

┌─1─┐
│ 1 │
└───┘

1 rows in set. Elapsed: 0.462 sec. Processed 1.28 thousand rows, 10.25 KB (2.77 thousand rows/s., 22.16 KB/s.) 

SELECT 1
FROM 
(
    SELECT 1
    FROM test
    WHERE x IN 
    (
        SELECT x
        FROM test
        ORDER BY x ASC
        LIMIT 10000000, 1
    )
)

┌─1─┐
│ 1 │
└───┘

1 rows in set. Elapsed: 0.406 sec. Processed 1.28 thousand rows, 10.25 KB (3.16 thousand rows/s., 25.24 KB/s.) 

SELECT 1
FROM 
(
    SELECT 1
    FROM 
    (
        SELECT 1
        FROM test
        WHERE x IN 
        (
            SELECT x
            FROM test
            ORDER BY x ASC
            LIMIT 10000000, 1
        )
    )
)

┌─1─┐
│ 1 │
└───┘

1 rows in set. Elapsed: 0.437 sec. Processed 1.28 thousand rows, 10.25 KB (2.93 thousand rows/s., 23.46 KB/s.) 
```

**Comment by filimonov** [comment]
https://github.com/ClickHouse/ClickHouse/pull/9261
